internal/http/download.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2024 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/k0sproject/k0s/pkg/build"
+)
+
+type DownloadOption func(*downloadOptions)
+
+// Downloads the contents of the given URL. Writes the HTTP response body to writer.
+func Download(ctx context.Context, url string, target io.Writer, options ...DownloadOption) (err error) {
+	opts := downloadOptions{}
+	for _, opt := range options {
+		opt(&opts)
+	}
+
+	// Prepare the client and the request.
+	client := http.Client{
+		Transport: &http.Transport{
+			// This is a one-shot HTTP client which should release resources immediately.
+			DisableKeepAlives: true,
+		},
+	}
```

Comment:
Maybe connect/read timeouts would be good?

Something like

```go
client := http.Client{
    Transport: &http.Transport{
        // Timeout for establishing a connection
        DialContext: (&net.Dialer{
            Timeout: opts.requestTimeout,
        }).DialContext,
        ResponseHeaderTimeout: opts.requestTimeout,
        DisableKeepAlives:     true,
        IdleConnTimeout: opts.requestTimeout,
    },
}
```

Reply:
I thought we could get away with a global external download timeout via the context. I've thought about cancelling stale downloads, but decided to do it in a follow-up PR: #5025.

Reply:
The full context timeout has to be quite long because of the file sizes for upgrades, so it would be nice to cancel early if the initial connection or response takes forever. We can do without, but it's so easy to add that it could be nice to add.

Reply:
The staleness timeout in #5025 would address this, I suppose. It's also sort of a time-to-first-byte timeout. I put it in a separate PR to make this one a bit smaller, and to be able to backport this PR separately without adding timeouts that weren't present in the grab version of the downloader. However, if you think it makes sense to have timeouts from the beginning, I'll add the commits here.